### ISSUE TITLE
Custom Data Training

### ISSUE THREAD

**Issue by KKaranKK** [clovaai/spade#2]
Thank you for the amazing work. The approach mentioned in published paper accurately addresses the problem of relationship between tokens (text detected by OCR). I want to train SPADE on Custom dataset. Could you please explain how to annotate data for training? 
- Most Importantly how to define field nodes (filled blue circles)

**Comment by whwang299** [comment]
Hi @KKaranKK 

You may change `field_representers` and `fields` (see below) https://github.com/clovaai/spade/blob/a85574ceaa00f1878a23754f283aa66bc2daf082/configs/cord.1.5layers.train.yaml#L35-L82


Internally, each field in `fields` is represented as learnable embedding. It is randomly initialized currently but you may use token embedding for better performance.
https://github.com/clovaai/spade/blob/a85574ceaa00f1878a23754f283aa66bc2daf082/spade/model/model_spade_graph_generator.py#L46-L48
https://github.com/clovaai/spade/blob/a85574ceaa00f1878a23754f283aa66bc2daf082/spade/model/model_spade_graph_generator.py#L224-L226


To make your own training set, you may check [CORD](https://github.com/clovaai/cord) dataset (and their [paper](https://openreview.net/pdf?id=SJl3z659UH))

 and how we have generate relation tags from them
https://github.com/clovaai/spade/blob/a85574ceaa00f1878a23754f283aa66bc2daf082/spade/model/data_utils.py#L516

Also check how we preprocess FUNSD dataset
https://github.com/clovaai/spade/blob/a85574ceaa00f1878a23754f283aa66bc2daf082/spade/model/data_utils.py#L882


Alternatively you can label relation between text segments directly but you may need to build your own labeling tools 
(or this [tool](https://github.com/allenai/pawls) may be useful)


Good luck!

Wonseok

**Comment by KKaranKK** [comment]
Thanks for the information.
Currently, I was trying to regenerate the results of ELB and ELK task over the FUNSD dataset for which I am struggling to download the pretrained bert-based-multilingual model using download_pretrained_models.py. Just need to confirm whether this project works on a non-cuda device or is it necessary to have a GPU?

**Comment by whwang299** [comment]
It is not necessary to have a gpu but the "training" will take a looooooooooong time (you may use "colab").
For the prediction, you can do that on cpu (it will take few seconds).

**Comment by KKaranKK** [comment]
For better understanding, I am training it on FUNSD dataset. While training on Colab I am encountering a memory issue. The error message states:
**i_sep = 1 / n_sep = 1
Fail to encode due to the memory limit.
The encoder output vectors set to zero.
Even single encoding faield!**  
The config used was -> **funsd.1.5layers.train.yaml**

**Comment by whwang299** [comment]
The error occurs because 16gb memory is not enough.

I can suggest three solutions.
(1) use gpu with at least 24gb memory.

(2) [modify the relative attention](https://github.com/clovaai/spade/blob/a85574ceaa00f1878a23754f283aa66bc2daf082/spade/model/model_2d_bert.py#L177-L194) (the relative attention used in SPADE costs a large memory. You may use `T5` style relative attention or [`BROS` style](https://arxiv.org/pdf/2108.04539.pdf)). 

(3) turn off relative attention by setting [`no_rel_attention: true`](https://github.com/clovaai/spade/blob/a85574ceaa00f1878a23754f283aa66bc2daf082/configs/funsd.1.5layers.train.yaml#L59) and uncomment [`absPos`](https://github.com/clovaai/spade/blob/a85574ceaa00f1878a23754f283aa66bc2daf082/configs/funsd.1.5layers.train.yaml#L20).
This way the model may cost less memory but there will be small drop in the performance.

Also, once you have succeeded running the training loop, please set [`toy_data: false`](https://github.com/clovaai/spade/blob/a85574ceaa00f1878a23754f283aa66bc2daf082/configs/funsd.1.5layers.train.yaml#L79).

**Comment by KKaranKK** [comment]
It worked, for the time being I am choosing option (3). Thanks for the work around. Moreover, It's more like a personal advice; I don't have much experience with BERT but as far as I understood your approach heavily depends on this as part of the Spatial Text Encoder. Would you think first understanding the original paper "Attention is all you need" will be helpful to do some changes with the approach?

**Comment by whwang299** [comment]
Yes, if you are not familiar with Transformer. After that you may read other Transformer papers using relative attention such as T5, Transformer-XL, BROS etc.. Good luck!
